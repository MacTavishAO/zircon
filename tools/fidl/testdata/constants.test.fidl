library values;

const uint8 UINT8 = 0b100;
const uint16 UINT16 = 0b100;
const uint32 UINT32 = 0b100;
const uint64 UINT64 = 0b100;
const int8 INT8 = 0b100;
const int16 INT16 = 0b100;
const int32 INT32 = 0b100;
const int64 INT64 = 0b100;
const float32 FLOAT32 = 3.14159;
const float64 FLOAT64 = 3.14159;
const bool BOOL = true;
const string STRING = "string";

enum EnumType : int32 {
    VALUE = 1;
};
const EnumType enumVal = EnumType.VALUE;
const int32 enumPrimitiveVal = EnumType.VALUE;

bits BitsType : uint32 {
    VALUE = 0x00000001;
    SECOND_VALUE = 0x80;
    THIRD_VALUE = 0x2;
};
const BitsType bitsVal = BitsType.VALUE;
const int32 bitsPrimitiveVal = BitsType.VALUE;

const BitsType orResult = BitsType.VALUE | BitsType.SECOND_VALUE | BitsType.THIRD_VALUE;
const int32 orResultPrimitiveVal = BitsType.VALUE | BitsType.SECOND_VALUE;

enum Enum {
    E = 0b10101010;
};

bits Bits {
    B = 0x8;
};

struct Struct {
    int64 int64_with_default = 007;
    string string_with_default = "stuff";
    bool bool_with_default = true;
    Enum enum_with_default = Enum.E;
    Bits bits_with_default = Bits.B;
};
