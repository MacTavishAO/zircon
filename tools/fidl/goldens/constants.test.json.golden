{
  "version": "0.0.1",
  "name": "values",
  "library_dependencies": [],
  "bits_declarations": [
    {
      "name": "values/BitsType",
      "location": {
        "filename": "constants.test.fidl",
        "line": 22,
        "column": 6,
        "length": 8
      },
      "type": {
        "kind": "primitive",
        "subtype": "uint32"
      },
      "mask": "131",
      "members": [
        {
          "name": "VALUE",
          "location": {
            "filename": "constants.test.fidl",
            "line": 23,
            "column": 5,
            "length": 5
          },
          "value": {
            "kind": "literal",
            "value": "1",
            "expression": "0x00000001",
            "literal": {
              "kind": "numeric",
              "value": "1",
              "expression": "0x00000001"
            }
          }
        },
        {
          "name": "SECOND_VALUE",
          "location": {
            "filename": "constants.test.fidl",
            "line": 24,
            "column": 5,
            "length": 12
          },
          "value": {
            "kind": "literal",
            "value": "128",
            "expression": "0x80",
            "literal": {
              "kind": "numeric",
              "value": "128",
              "expression": "0x80"
            }
          }
        },
        {
          "name": "THIRD_VALUE",
          "location": {
            "filename": "constants.test.fidl",
            "line": 25,
            "column": 5,
            "length": 11
          },
          "value": {
            "kind": "literal",
            "value": "2",
            "expression": "0x2",
            "literal": {
              "kind": "numeric",
              "value": "2",
              "expression": "0x2"
            }
          }
        }
      ],
      "strict": true
    },
    {
      "name": "values/Bits",
      "location": {
        "filename": "constants.test.fidl",
        "line": 37,
        "column": 6,
        "length": 4
      },
      "type": {
        "kind": "primitive",
        "subtype": "uint32"
      },
      "mask": "8",
      "members": [
        {
          "name": "B",
          "location": {
            "filename": "constants.test.fidl",
            "line": 38,
            "column": 5,
            "length": 1
          },
          "value": {
            "kind": "literal",
            "value": "8",
            "expression": "0x8",
            "literal": {
              "kind": "numeric",
              "value": "8",
              "expression": "0x8"
            }
          }
        }
      ],
      "strict": true
    }
  ],
  "const_declarations": [
    {
      "name": "values/UINT8",
      "location": {
        "filename": "constants.test.fidl",
        "line": 3,
        "column": 13,
        "length": 5
      },
      "type": {
        "kind": "primitive",
        "subtype": "uint8"
      },
      "value": {
        "kind": "literal",
        "value": "4",
        "expression": "0b100",
        "literal": {
          "kind": "numeric",
          "value": "4",
          "expression": "0b100"
        }
      }
    },
    {
      "name": "values/UINT16",
      "location": {
        "filename": "constants.test.fidl",
        "line": 4,
        "column": 14,
        "length": 6
      },
      "type": {
        "kind": "primitive",
        "subtype": "uint16"
      },
      "value": {
        "kind": "literal",
        "value": "4",
        "expression": "0b100",
        "literal": {
          "kind": "numeric",
          "value": "4",
          "expression": "0b100"
        }
      }
    },
    {
      "name": "values/UINT32",
      "location": {
        "filename": "constants.test.fidl",
        "line": 5,
        "column": 14,
        "length": 6
      },
      "type": {
        "kind": "primitive",
        "subtype": "uint32"
      },
      "value": {
        "kind": "literal",
        "value": "4",
        "expression": "0b100",
        "literal": {
          "kind": "numeric",
          "value": "4",
          "expression": "0b100"
        }
      }
    },
    {
      "name": "values/UINT64",
      "location": {
        "filename": "constants.test.fidl",
        "line": 6,
        "column": 14,
        "length": 6
      },
      "type": {
        "kind": "primitive",
        "subtype": "uint64"
      },
      "value": {
        "kind": "literal",
        "value": "4",
        "expression": "0b100",
        "literal": {
          "kind": "numeric",
          "value": "4",
          "expression": "0b100"
        }
      }
    },
    {
      "name": "values/INT8",
      "location": {
        "filename": "constants.test.fidl",
        "line": 7,
        "column": 12,
        "length": 4
      },
      "type": {
        "kind": "primitive",
        "subtype": "int8"
      },
      "value": {
        "kind": "literal",
        "value": "4",
        "expression": "0b100",
        "literal": {
          "kind": "numeric",
          "value": "4",
          "expression": "0b100"
        }
      }
    },
    {
      "name": "values/INT16",
      "location": {
        "filename": "constants.test.fidl",
        "line": 8,
        "column": 13,
        "length": 5
      },
      "type": {
        "kind": "primitive",
        "subtype": "int16"
      },
      "value": {
        "kind": "literal",
        "value": "4",
        "expression": "0b100",
        "literal": {
          "kind": "numeric",
          "value": "4",
          "expression": "0b100"
        }
      }
    },
    {
      "name": "values/INT32",
      "location": {
        "filename": "constants.test.fidl",
        "line": 9,
        "column": 13,
        "length": 5
      },
      "type": {
        "kind": "primitive",
        "subtype": "int32"
      },
      "value": {
        "kind": "literal",
        "value": "4",
        "expression": "0b100",
        "literal": {
          "kind": "numeric",
          "value": "4",
          "expression": "0b100"
        }
      }
    },
    {
      "name": "values/INT64",
      "location": {
        "filename": "constants.test.fidl",
        "line": 10,
        "column": 13,
        "length": 5
      },
      "type": {
        "kind": "primitive",
        "subtype": "int64"
      },
      "value": {
        "kind": "literal",
        "value": "4",
        "expression": "0b100",
        "literal": {
          "kind": "numeric",
          "value": "4",
          "expression": "0b100"
        }
      }
    },
    {
      "name": "values/FLOAT32",
      "location": {
        "filename": "constants.test.fidl",
        "line": 11,
        "column": 15,
        "length": 7
      },
      "type": {
        "kind": "primitive",
        "subtype": "float32"
      },
      "value": {
        "kind": "literal",
        "value": "3.14159",
        "expression": "3.14159",
        "literal": {
          "kind": "numeric",
          "value": "3.14159",
          "expression": "3.14159"
        }
      }
    },
    {
      "name": "values/FLOAT64",
      "location": {
        "filename": "constants.test.fidl",
        "line": 12,
        "column": 15,
        "length": 7
      },
      "type": {
        "kind": "primitive",
        "subtype": "float64"
      },
      "value": {
        "kind": "literal",
        "value": "3.14159",
        "expression": "3.14159",
        "literal": {
          "kind": "numeric",
          "value": "3.14159",
          "expression": "3.14159"
        }
      }
    },
    {
      "name": "values/BOOL",
      "location": {
        "filename": "constants.test.fidl",
        "line": 13,
        "column": 12,
        "length": 4
      },
      "type": {
        "kind": "primitive",
        "subtype": "bool"
      },
      "value": {
        "kind": "literal",
        "value": "true",
        "expression": "true",
        "literal": {
          "kind": "true",
          "value": "true",
          "expression": "true"
        }
      }
    },
    {
      "name": "values/STRING",
      "location": {
        "filename": "constants.test.fidl",
        "line": 14,
        "column": 14,
        "length": 6
      },
      "type": {
        "kind": "string",
        "nullable": false
      },
      "value": {
        "kind": "literal",
        "value": "string",
        "expression": "\"string\"",
        "literal": {
          "kind": "string",
          "value": "string",
          "expression": "\"string\""
        }
      }
    },
    {
      "name": "values/enumVal",
      "location": {
        "filename": "constants.test.fidl",
        "line": 19,
        "column": 16,
        "length": 7
      },
      "type": {
        "kind": "identifier",
        "identifier": "values/EnumType",
        "nullable": false
      },
      "value": {
        "kind": "identifier",
        "value": "1",
        "expression": "EnumType.VALUE",
        "identifier": "values/EnumType.VALUE"
      }
    },
    {
      "name": "values/enumPrimitiveVal",
      "location": {
        "filename": "constants.test.fidl",
        "line": 20,
        "column": 13,
        "length": 16
      },
      "type": {
        "kind": "primitive",
        "subtype": "int32"
      },
      "value": {
        "kind": "identifier",
        "value": "1",
        "expression": "EnumType.VALUE",
        "identifier": "values/EnumType.VALUE"
      }
    },
    {
      "name": "values/bitsVal",
      "location": {
        "filename": "constants.test.fidl",
        "line": 27,
        "column": 16,
        "length": 7
      },
      "type": {
        "kind": "identifier",
        "identifier": "values/BitsType",
        "nullable": false
      },
      "value": {
        "kind": "identifier",
        "value": "1",
        "expression": "BitsType.VALUE",
        "identifier": "values/BitsType.VALUE"
      }
    },
    {
      "name": "values/bitsPrimitiveVal",
      "location": {
        "filename": "constants.test.fidl",
        "line": 28,
        "column": 13,
        "length": 16
      },
      "type": {
        "kind": "primitive",
        "subtype": "int32"
      },
      "value": {
        "kind": "identifier",
        "value": "1",
        "expression": "BitsType.VALUE",
        "identifier": "values/BitsType.VALUE"
      }
    },
    {
      "name": "values/orResult",
      "location": {
        "filename": "constants.test.fidl",
        "line": 30,
        "column": 16,
        "length": 8
      },
      "type": {
        "kind": "identifier",
        "identifier": "values/BitsType",
        "nullable": false
      },
      "value": {
        "kind": "binary_operator",
        "value": "131",
        "expression": "BitsType.VALUE | BitsType.SECOND_VALUE | BitsType.THIRD_VALUE"
      }
    },
    {
      "name": "values/orResultPrimitiveVal",
      "location": {
        "filename": "constants.test.fidl",
        "line": 31,
        "column": 13,
        "length": 20
      },
      "type": {
        "kind": "primitive",
        "subtype": "int32"
      },
      "value": {
        "kind": "binary_operator",
        "value": "129",
        "expression": "BitsType.VALUE | BitsType.SECOND_VALUE"
      }
    }
  ],
  "enum_declarations": [
    {
      "name": "values/EnumType",
      "location": {
        "filename": "constants.test.fidl",
        "line": 16,
        "column": 6,
        "length": 8
      },
      "type": "int32",
      "members": [
        {
          "name": "VALUE",
          "location": {
            "filename": "constants.test.fidl",
            "line": 17,
            "column": 5,
            "length": 5
          },
          "value": {
            "kind": "literal",
            "value": "1",
            "expression": "1",
            "literal": {
              "kind": "numeric",
              "value": "1",
              "expression": "1"
            }
          }
        }
      ],
      "strict": true
    },
    {
      "name": "values/Enum",
      "location": {
        "filename": "constants.test.fidl",
        "line": 33,
        "column": 6,
        "length": 4
      },
      "type": "uint32",
      "members": [
        {
          "name": "E",
          "location": {
            "filename": "constants.test.fidl",
            "line": 34,
            "column": 5,
            "length": 1
          },
          "value": {
            "kind": "literal",
            "value": "170",
            "expression": "0b10101010",
            "literal": {
              "kind": "numeric",
              "value": "170",
              "expression": "0b10101010"
            }
          }
        }
      ],
      "strict": true
    }
  ],
  "experimental_resource_declarations": [],
  "interface_declarations": [],
  "service_declarations": [],
  "struct_declarations": [
    {
      "name": "values/Struct",
      "location": {
        "filename": "constants.test.fidl",
        "line": 41,
        "column": 8,
        "length": 6
      },
      "anonymous": false,
      "members": [
        {
          "type": {
            "kind": "primitive",
            "subtype": "int64"
          },
          "name": "int64_with_default",
          "location": {
            "filename": "constants.test.fidl",
            "line": 42,
            "column": 11,
            "length": 18
          },
          "maybe_default_value": {
            "kind": "literal",
            "value": "7",
            "expression": "007",
            "literal": {
              "kind": "numeric",
              "value": "7",
              "expression": "007"
            }
          },
          "field_shape_v1": {
            "offset": 0,
            "padding": 0
          }
        },
        {
          "type": {
            "kind": "string",
            "nullable": false
          },
          "name": "string_with_default",
          "location": {
            "filename": "constants.test.fidl",
            "line": 43,
            "column": 12,
            "length": 19
          },
          "maybe_default_value": {
            "kind": "literal",
            "value": "stuff",
            "expression": "\"stuff\"",
            "literal": {
              "kind": "string",
              "value": "stuff",
              "expression": "\"stuff\""
            }
          },
          "field_shape_v1": {
            "offset": 8,
            "padding": 0
          }
        },
        {
          "type": {
            "kind": "primitive",
            "subtype": "bool"
          },
          "name": "bool_with_default",
          "location": {
            "filename": "constants.test.fidl",
            "line": 44,
            "column": 10,
            "length": 17
          },
          "maybe_default_value": {
            "kind": "literal",
            "value": "true",
            "expression": "true",
            "literal": {
              "kind": "true",
              "value": "true",
              "expression": "true"
            }
          },
          "field_shape_v1": {
            "offset": 24,
            "padding": 3
          }
        },
        {
          "type": {
            "kind": "identifier",
            "identifier": "values/Enum",
            "nullable": false
          },
          "name": "enum_with_default",
          "location": {
            "filename": "constants.test.fidl",
            "line": 45,
            "column": 10,
            "length": 17
          },
          "maybe_default_value": {
            "kind": "identifier",
            "value": "170",
            "expression": "Enum.E",
            "identifier": "values/Enum.E"
          },
          "field_shape_v1": {
            "offset": 28,
            "padding": 0
          }
        },
        {
          "type": {
            "kind": "identifier",
            "identifier": "values/Bits",
            "nullable": false
          },
          "name": "bits_with_default",
          "location": {
            "filename": "constants.test.fidl",
            "line": 46,
            "column": 10,
            "length": 17
          },
          "maybe_default_value": {
            "kind": "identifier",
            "value": "8",
            "expression": "Bits.B",
            "identifier": "values/Bits.B"
          },
          "field_shape_v1": {
            "offset": 32,
            "padding": 4
          }
        }
      ],
      "type_shape_v1": {
        "inline_size": 40,
        "alignment": 8,
        "depth": 1,
        "max_handles": 0,
        "max_out_of_line": 4294967295,
        "has_padding": true,
        "has_flexible_envelope": false,
        "is_resource": false
      }
    }
  ],
  "table_declarations": [],
  "union_declarations": [],
  "type_alias_declarations": [],
  "declaration_order": [
    "values/UINT8",
    "values/UINT64",
    "values/UINT32",
    "values/UINT16",
    "values/STRING",
    "values/INT8",
    "values/INT64",
    "values/INT32",
    "values/INT16",
    "values/FLOAT64",
    "values/FLOAT32",
    "values/EnumType",
    "values/enumVal",
    "values/enumPrimitiveVal",
    "values/Enum",
    "values/BitsType",
    "values/orResultPrimitiveVal",
    "values/orResult",
    "values/bitsVal",
    "values/bitsPrimitiveVal",
    "values/Bits",
    "values/Struct",
    "values/BOOL"
  ],
  "declarations": {
    "values/BitsType": "bits",
    "values/Bits": "bits",
    "values/UINT8": "const",
    "values/UINT16": "const",
    "values/UINT32": "const",
    "values/UINT64": "const",
    "values/INT8": "const",
    "values/INT16": "const",
    "values/INT32": "const",
    "values/INT64": "const",
    "values/FLOAT32": "const",
    "values/FLOAT64": "const",
    "values/BOOL": "const",
    "values/STRING": "const",
    "values/enumVal": "const",
    "values/enumPrimitiveVal": "const",
    "values/bitsVal": "const",
    "values/bitsPrimitiveVal": "const",
    "values/orResult": "const",
    "values/orResultPrimitiveVal": "const",
    "values/EnumType": "enum",
    "values/Enum": "enum",
    "values/Struct": "struct"
  }
}
